docs_src/design/adr/0006-Service-Registry.md

```diff
@@ -0,0 +1,192 @@
+of EdgeX as well as versions through the Fuji release all relied on a bootstrapping service called core-config-seed which was
+responsible for seeding the configuration of all of the core and support services into Consul prior to any of the services being
+started.
+
+This release actually preceded usage of TOML for configuration files, and instead just used a flat key/value format,
+with keys converted from legacy Java property names (e.g. meta.db.device.url ) to Camel[Pascal]/Case (e.g. MetaDeviceServiceURL).
+
+I chose the config key mentioned above on purpose:
+
+```
+MetaDeviceURL = "http://edgex-core-metadata:48081/api/v1/device"
+```
+
+Not only did this config key provide the address of core metadata, it also provided the path of a specific REST endpoint. In later releases
+of EdgeX, the address of the service and the specific endpoint paths were de-coupled. Instead of following the Service Name design (which
+was finalized two months earlier), the initial implementation followed the legacy Java implementation and initialized its service clients
+for each required REST endpoint (belonging to another EdgeX service) directly from the associated *URL config key read from Consul (if
+enabled) or directly from the configuration file.
+
+The shared client initialization code also created an Endpoint monitor goroutine and passed it a go channel channel used by the service
+to receive updates to the REST API endpoint URL. This monitor goroutine effectively polled Consul every 15s (this became configurable in
+later versions) for the client's service address and if a change was detected, would write the updated endpoint URL to the given channel,
+effectively ensuring that the service started using the new URL.
+
+It wasn't till late in the Geneva development cycle that I noticed log messages which made me aware of the fact that every one of our
+services was making a REST call to check the address of a service endpoint every 15s, for **every** REST endpoint it used! An issue was
+filed (https://github.com/edgexfoundry/edgex-go/issues/2594), and the client monitoring was removed as part of the Geneva 1.2.1 release.
+
+## Problem Statement
+The fundamental problem with the existing implementations (as decribed above), is that there is too much duplication of configuration across
+services. For instance, Core Data's service port can easily be changed by passing the environment variable SERVICE_PORT to the service on
+startup. This overrides the configuration read from the configuration provider, and will cause Core Data to listen on the new port,
+however it has no impact on any services which use Core Data, as the client config for each is read from the configuration provider (excluding
+security-proxy-setup).
+
+This means in order to change a service port, environment variable overrides need to set for every client service as well as security-proxy-setup
+(if required).
```

Comment:
For this to work the `CLIENTS_COREDATA_PORT` environment variable would also need to be changed for each service. In docker we add it to the common.env file and re-generate the the compose files, which sets the override on each service. For snap is there a single environment for all the service or does each have it own like in docker?

Reply:
Mentioned `CLIENTS_COREDATA_PORT` in the text.

As for whether snaps have the ability to define a single environment variable for all services...  Yes, environment variables can be defined at buildtime (i.e. in snapcraft.yaml) which apply to all services. At runtime, this is up to the snap developer. In fact this is actually implemented in the edgexfoundry snap for `EDGEX_SECURITY_SECRET_STORE` variable. So in theory, it *could* also be implemented for client addressing if/when the registry was not in use.